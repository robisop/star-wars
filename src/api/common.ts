export interface SwapiListResponse {
  count: number;
  next: string | null;
  previous: string | null;
  results: ListItemData[];
}

export interface ListItemData {
  id: string;
  title: string;
  subtitle: string;
}
