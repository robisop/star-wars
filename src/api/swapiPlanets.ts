import {formatNumberString} from 'helpers/formatters';

import {SwapiListResponse} from './common';

export interface SwapiPlanets {
  count: number;
  next: string | null;
  previous: string | null;
  results: SwapiPlanet[];
}

export interface SwapiPlanet {
  name: string;
  rotation_period: string;
  orbital_period: string;
  diameter: string;
  climate: string;
  gravity: string;
  terrain: string;
  surface_water: string;
  population: string;
  residents: string[];
  films: string[];
  created: string;
  edited: string;
  url: string;
}

export const parseSwapiPlanets = (data: SwapiPlanets): SwapiListResponse => {
  return {
    count: data.count,
    next: data.next,
    previous: data.previous,
    results: data.results.map(planet => ({
      id: planet.url,
      title: planet.name,
      subtitle: `population: ${formatNumberString(
        planet.population,
      )} | climate: ${planet.climate} | terrain: ${
        planet.terrain
      } | diameter: ${formatNumberString(planet.diameter)} | gravity: ${
        planet.gravity
      }`,
    })),
  };
};

export const mockSwapiPlanets: SwapiPlanets = {
  count: 3,
  next: null,
  previous: null,
  results: [
    {
      name: 'Tatooine',
      rotation_period: '23',
      orbital_period: '304',
      diameter: '10465',
      climate: 'arid',
      gravity: '1 standard',
      terrain: 'desert',
      surface_water: '1',
      population: '200000',
      residents: [
        'https://swapi.dev/api/people/1/',
        'https://swapi.dev/api/people/2/',
        'https://swapi.dev/api/people/4/',
        'https://swapi.dev/api/people/6/',
        'https://swapi.dev/api/people/7/',
        'https://swapi.dev/api/people/8/',
        'https://swapi.dev/api/people/9/',
        'https://swapi.dev/api/people/11/',
        'https://swapi.dev/api/people/43/',
        'https://swapi.dev/api/people/62/',
      ],
      films: [
        'https://swapi.dev/api/films/1/',
        'https://swapi.dev/api/films/3/',
        'https://swapi.dev/api/films/4/',
        'https://swapi.dev/api/films/5/',
        'https://swapi.dev/api/films/6/',
      ],
      created: '2014-12-09T13:50:49.641000Z',
      edited: '2014-12-20T20:58:18.411000Z',
      url: 'https://swapi.dev/api/planets/1/',
    },
    {
      name: 'Alderaan',
      rotation_period: '24',
      orbital_period: '364',
      diameter: '12500',
      climate: 'temperate',
      gravity: '1 standard',
      terrain: 'grasslands, mountains',
      surface_water: '40',
      population: '2000000000',
      residents: [
        'https://swapi.dev/api/people/5/',
        'https://swapi.dev/api/people/68/',
        'https://swapi.dev/api/people/81/',
      ],
      films: [
        'https://swapi.dev/api/films/1/',
        'https://swapi.dev/api/films/6/',
      ],
      created: '2014-12-10T11:35:48.479000Z',
      edited: '2014-12-20T20:58:18.420000Z',
      url: 'https://swapi.dev/api/planets/2/',
    },
    {
      name: 'Yavin IV',
      rotation_period: '24',
      orbital_period: '4818',
      diameter: '10200',
      climate: 'temperate, tropical',
      gravity: '1 standard',
      terrain: 'jungle, rainforests',
      surface_water: '8',
      population: '1000',
      residents: [],
      films: ['https://swapi.dev/api/films/1/'],
      created: '2014-12-10T11:37:19.144000Z',
      edited: '2014-12-20T20:58:18.421000Z',
      url: 'https://swapi.dev/api/planets/3/',
    },
  ],
};
