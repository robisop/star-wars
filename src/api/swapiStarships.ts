import {formatNumberString} from 'helpers/formatters';

import {SwapiListResponse} from './common';

export interface SwapiStarships {
  count: number;
  next: string | null;
  previous: string | null;
  results: SwapiStarship[];
}

export interface SwapiStarship {
  name: string;
  model: string;
  manufacturer: string;
  cost_in_credits: string;
  length: string;
  max_atmosphering_speed: string;
  crew: string;
  passengers: string;
  cargo_capacity: string;
  consumables: string;
  hyperdrive_rating: string;
  MGLT: string;
  starship_class: string;
  pilots: string[];
  films: string[];
  created: string;
  edited: string;
  url: string;
}

export const parseSwapiStarships = (
  data: SwapiStarships,
): SwapiListResponse => {
  return {
    count: data.count,
    next: data.next,
    previous: data.previous,
    results: data.results.map(starship => ({
      id: starship.url,
      title: starship.name,
      subtitle: `model: ${starship.model} | manufacturer: ${
        starship.manufacturer
      } | cost: ${formatNumberString(
        starship.cost_in_credits,
      )} credits | length: ${formatNumberString(
        starship.length,
      )} | max speed: ${formatNumberString(
        starship.max_atmosphering_speed,
      )} | crew ${formatNumberString(
        starship.crew,
      )} | passengers: ${formatNumberString(
        starship.passengers,
      )} | cargo capacity: ${formatNumberString(
        starship.cargo_capacity,
      )} | hyperdrive rating: ${starship.hyperdrive_rating} | starship class: ${
        starship.starship_class
      }`,
    })),
  };
};

export const mockSwapiStarships: SwapiStarships = {
  count: 3,
  next: null,
  previous: null,
  results: [
    {
      name: 'Star Destroyer',
      model: 'Imperial I-class Star Destroyer',
      manufacturer: 'Kuat Drive Yards',
      cost_in_credits: '150000000',
      length: '1,600',
      max_atmosphering_speed: '975',
      crew: '47,060',
      passengers: 'n/a',
      cargo_capacity: '36000000',
      consumables: '2 years',
      hyperdrive_rating: '2.0',
      MGLT: '60',
      starship_class: 'Star Destroyer',
      pilots: [],
      films: [
        'https://swapi.dev/api/films/1/',
        'https://swapi.dev/api/films/2/',
        'https://swapi.dev/api/films/3/',
      ],
      created: '2014-12-10T15:08:19.848000Z',
      edited: '2014-12-20T21:23:49.870000Z',
      url: 'https://swapi.dev/api/starships/3/',
    },
    {
      name: 'Sentinel-class landing craft',
      model: 'Sentinel-class landing craft',
      manufacturer: 'Sienar Fleet Systems, Cyngus Spaceworks',
      cost_in_credits: '240000',
      length: '38',
      max_atmosphering_speed: '1000',
      crew: '5',
      passengers: '75',
      cargo_capacity: '180000',
      consumables: '1 month',
      hyperdrive_rating: '1.0',
      MGLT: '70',
      starship_class: 'landing craft',
      pilots: [],
      films: ['https://swapi.dev/api/films/1/'],
      created: '2014-12-10T15:48:00.586000Z',
      edited: '2014-12-20T21:23:49.873000Z',
      url: 'https://swapi.dev/api/starships/5/',
    },
    {
      name: 'Death Star',
      model: 'DS-1 Orbital Battle Station',
      manufacturer:
        'Imperial Department of Military Research, Sienar Fleet Systems',
      cost_in_credits: '1000000000000',
      length: '120000',
      max_atmosphering_speed: 'n/a',
      crew: '342,953',
      passengers: '843,342',
      cargo_capacity: '1000000000000',
      consumables: '3 years',
      hyperdrive_rating: '4.0',
      MGLT: '10',
      starship_class: 'Deep Space Mobile Battlestation',
      pilots: [],
      films: ['https://swapi.dev/api/films/1/'],
      created: '2014-12-10T16:36:50.509000Z',
      edited: '2014-12-20T21:26:24.783000Z',
      url: 'https://swapi.dev/api/starships/9/',
    },
    {
      name: 'Millennium Falcon',
      model: 'YT-1300 light freighter',
      manufacturer: 'Corellian Engineering Corporation',
      cost_in_credits: '100000',
      length: '34.37',
      max_atmosphering_speed: '1050',
      crew: '4',
      passengers: '6',
      cargo_capacity: '100000',
      consumables: '2 months',
      hyperdrive_rating: '0.5',
      MGLT: '75',
      starship_class: 'Light freighter',
      pilots: [
        'https://swapi.dev/api/people/13/',
        'https://swapi.dev/api/people/14/',
        'https://swapi.dev/api/people/25/',
        'https://swapi.dev/api/people/31/',
      ],
      films: [
        'https://swapi.dev/api/films/1/',
        'https://swapi.dev/api/films/2/',
        'https://swapi.dev/api/films/3/',
      ],
      created: '2014-12-10T16:59:45.094000Z',
      edited: '2014-12-20T21:23:49.880000Z',
      url: 'https://swapi.dev/api/starships/10/',
    },
  ],
};
