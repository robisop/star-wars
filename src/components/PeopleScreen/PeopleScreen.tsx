import {ListView} from 'components/common/ListView';
import {usePeople} from 'hooks/usePeople';

export default function PeopleScreen() {
  const {
    data,
    isLoading,
    isError,
    fetchNextPage,
    hasNextPage,
    isFetchingNextPage,
  } = usePeople();

  return (
    <ListView
      data={data?.pages.map(page => page.results).flat()}
      hasNextPage={hasNextPage}
      isLoading={isLoading}
      isError={isError}
      isFetchingNextPage={isFetchingNextPage}
      fetchNextPage={fetchNextPage}
    />
  );
}
