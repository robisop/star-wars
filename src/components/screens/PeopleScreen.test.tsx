import {mockSwapiPeople} from 'api/swapiPeople';
import renderer from 'react-test-renderer';

import PeopleScreen from './PeopleScreen';

jest.mock('hooks/usePeople', () => ({
  usePeople: jest.fn(() => ({
    data: {
      pages: [
        jest.requireActual('api/swapiPeople').parseSwapiPeople(mockSwapiPeople),
      ],
    },
    isLoading: false,
    isError: false,
    fetchNextPage: jest.fn(),
    hasNextPage: false,
    isFetchingNextPage: false,
  })),
}));

describe('<PeopleScreen />', () => {
  it('renders correctly', () => {
    const component = renderer.create(<PeopleScreen />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
