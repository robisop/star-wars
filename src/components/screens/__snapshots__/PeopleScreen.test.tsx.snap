// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PeopleScreen /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    data={
      [
        {
          "id": "https://swapi.dev/api/people/1/",
          "subtitle": "gender: male | height: 172 | hair: blond | eyes: blue",
          "title": "Luke Skywalker",
        },
        {
          "id": "https://swapi.dev/api/people/2/",
          "subtitle": "gender: n/a | height: 167 | hair: n/a | eyes: yellow",
          "title": "C-3PO",
        },
        {
          "id": "https://swapi.dev/api/people/3/",
          "subtitle": "gender: n/a | height: 96 | hair: n/a | eyes: red",
          "title": "R2-D2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 10,
              "minWidth": "100%",
              "paddingBottom": 10,
              "paddingLeft": 30,
              "paddingRight": 10,
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              Luke Skywalker
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 10,
              }
            }
          >
            <Text>
              gender: male | height: 172 | hair: blond | eyes: blue
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 10,
              "minWidth": "100%",
              "paddingBottom": 10,
              "paddingLeft": 30,
              "paddingRight": 10,
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              C-3PO
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 10,
              }
            }
          >
            <Text>
              gender: n/a | height: 167 | hair: n/a | eyes: yellow
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 10,
              "minWidth": "100%",
              "paddingBottom": 10,
              "paddingLeft": 30,
              "paddingRight": 10,
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              R2-D2
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 10,
              }
            }
          >
            <Text>
              gender: n/a | height: 96 | hair: n/a | eyes: red
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
